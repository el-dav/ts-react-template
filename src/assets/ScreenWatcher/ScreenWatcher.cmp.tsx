import React, { Component } from 'react';
import _forEach from 'lodash/forEach';
import _kebabCase from 'lodash/kebabCase';
import _memoize from 'lodash/memoize';
import _throttle from 'lodash/throttle';
import styled from 'react-emotion';

import { Props, ScreenData } from './ScreenWatcher.typ';

const getScreenClassFunc = (screenData: ScreenData) => {
  let className = '';

  _forEach(screenData, (value, key) => {
    if (value && key !== 'width' && key !== 'height') {
      const formattedKey = _kebabCase(key);
      className += ` screen-${formattedKey}`;
    }
  });

  return className.trim();
};
const getScreenClass = _memoize(getScreenClassFunc);

const Wrapper = styled('div')`
  width: 100%, auto;
  height: 100%;
`;

interface ResizeEvent {
  target: {
    innerWidth: number;
    innerHeight: number;
  };
}

class ScreenWatcher extends Component<Props> {
  constructor(props: Props) {
    super(props);

    this.updateDimensions = this.updateDimensions.bind(this);
  }

  public componentDidMount() {
    window.addEventListener(
      'resize',
      _throttle(this.updateDimensions.bind(this), 100)
    );
    this.props.screenResize(window.innerWidth, window.innerHeight);
  }

  public componentWillUnmount() {
    window.removeEventListener(
      'resize',
      _throttle(this.updateDimensions.bind(this), 100)
    );
  }

  public updateDimensions(e: ResizeEvent) {
    this.props.screenResize(e.target.innerWidth, e.target.innerHeight);
  }

  public render() {
    const { children, screenData, className, ...rest } = this.props;
    return (
      <Wrapper
        {...rest}
        className={`screen-watcher ${this.props.className ||
          ''} ${getScreenClass(screenData)}`}
      >
        {this.props.children}
      </Wrapper>
    );
  }
}

export default ScreenWatcher;
