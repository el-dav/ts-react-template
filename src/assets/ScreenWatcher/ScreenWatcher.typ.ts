import { State as ScreenData } from 'ducks/screen/reducers';

export interface OwnProps {
  className?: string;
}

export interface StateProps {
  className?: string;
  screenData: ScreenData;
}

export interface DispatchProps {
  screenResize: (width: number, height: number) => void;
}

export interface Props extends StateProps, DispatchProps {}
