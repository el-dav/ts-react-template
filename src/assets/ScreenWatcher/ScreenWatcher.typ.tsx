export interface ScreenData {
  xs: boolean;
  gtXs: boolean;
  sm: boolean;
  gtSm: boolean;
  md: boolean;
  gtMd: boolean;
  lg: boolean;
  gtLg: boolean;
  xl: boolean;
  width: number;
  height: number;
}

export interface OwnProps {
  className?: string;
}

export interface StateProps {
  className?: string;
  screenData: ScreenData;
}

export interface DispatchProps {
  screenResize: (width: number, height: number) => void;
}

export interface Props extends StateProps, DispatchProps {}
