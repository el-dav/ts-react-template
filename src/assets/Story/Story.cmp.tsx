import React from 'react';

import Theme from '~/assets/Theme';
import Store from '~/assets/Store';

interface Props {
  isConnected?: boolean;
}

const Story: React.SFC<Props> = ({ children, isConnected }) =>
  isConnected ? (
    <Store>
      <Theme>{children}</Theme>
    </Store>
  ) : (
    <Theme>{children}</Theme>
  );

export default Story;
