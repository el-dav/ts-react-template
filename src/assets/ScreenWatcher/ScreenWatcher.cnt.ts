import { connect } from 'react-redux';
import { Dispatch } from 'redux';

import { AppState } from 'ducks/ducks';
import { selectScreenData } from 'ducks/screen/selectors';
import { screenResize } from 'ducks/screen/actions';

import ScreenWatcher from './ScreenWatcher.cmp';

import { StateProps, DispatchProps, OwnProps } from './ScreenWatcher.typ';

const mapStateToProps = (state: AppState, ownProps: OwnProps): StateProps => ({
  className: ownProps.className,
  screenData: selectScreenData(state)
});

const mapDispatchToProps = (dispatch: Dispatch): DispatchProps => ({
  screenResize: (width: number, height: number) => {
    dispatch(screenResize(width, height));
  }
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ScreenWatcher);
