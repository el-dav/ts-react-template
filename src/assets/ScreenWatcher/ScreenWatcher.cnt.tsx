import { connect } from 'react-redux';

import { State } from 'ducks/ducks';
import { selectScreenData } from 'ducks/screen/selectors';
import { screenResize } from 'ducks/screen/actions';

import ScreenWatcher from './ScreenWatcher.cmp';

import { StateProps, DispatchProps, OwnProps } from './ScreenWatcher.typ';

const mapStateToProps = (state: State, ownProps: OwnProps): StateProps => ({
  className: ownProps.className,
  screenData: selectScreenData(state)
});

const mapDispatchToProps = (dispatch): DispatchProps => ({
  screenResize: (width, height) => {
    dispatch(screenResize(width, height));
  }
});

export default connect(mapStateToProps, mapDispatchToProps)(ScreenWatcher);
