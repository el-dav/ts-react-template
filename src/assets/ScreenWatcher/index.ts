import ScreenWatcher from './ScreenWatcher.cnt';
export default ScreenWatcher;
