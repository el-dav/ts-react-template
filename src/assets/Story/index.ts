import Story from './Story.cmp';

export default Story;
