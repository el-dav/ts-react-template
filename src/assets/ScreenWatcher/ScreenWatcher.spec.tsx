import React from 'react';
import renderer from 'react-test-renderer';

import Story from 'assets/Story';
import { SIZES } from 'ducks/screen/config';

import ScreenWatcherCmp from './ScreenWatcher.cmp';
import ScreenWatcherCnt from './ScreenWatcher.cnt';
import { ScreenData } from './ScreenWatcher.typ';

const xsScreenData = {
  width: SIZES.SM - 1,
  height: 500,
  xs: true,
  gtXs: false,
  sm: false,
  gtSm: false,
  md: false,
  gtMd: false,
  lg: false,
  gtLg: false,
  xl: false
};
const smScreenData = {
  ...xsScreenData,
  width: SIZES.MD - 1,
  xs: false,
  gtXs: true,
  sm: true
};
const mdScreenData = {
  ...smScreenData,
  width: SIZES.LG - 1,
  sm: false,
  gtSm: true,
  md: true
};
const lgScreenData = {
  ...mdScreenData,
  width: SIZES.XL - 1,
  md: false,
  gtMd: true,
  lg: true
};
const xlScreenData = {
  ...lgScreenData,
  width: SIZES.XL + 1,
  lg: false,
  gtLg: true,
  xl: true
};

const screenResize = (width: number, height: number) => undefined;

const testSize = (size: string, screenData: ScreenData) => {
  it(`renders the Component correctly for ${size} size`, () => {
    const tree = renderer
      .create(
        <Story isConnected>
          <ScreenWatcherCmp
            screenData={screenData}
            screenResize={screenResize}
          />
        </Story>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
};

describe('ScreenWatcher', () => {
  testSize('xs', xsScreenData);
  testSize('sm', smScreenData);
  testSize('md', mdScreenData);
  testSize('lg', lgScreenData);
  testSize('xl', xlScreenData);

  it('renders the Container correctly', () => {
    const tree = renderer
      .create(
        <Story isConnected>
          <ScreenWatcherCnt />
        </Story>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
