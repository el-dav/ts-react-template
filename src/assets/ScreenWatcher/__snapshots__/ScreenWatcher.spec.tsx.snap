// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScreenWatcher renders the Component correctly for lg size 1`] = `
<div
  className="screen-watcher screen-gt-xs screen-gt-sm screen-gt-md screen-lg css-1ow008w"
/>
`;

exports[`ScreenWatcher renders the Component correctly for md size 1`] = `
<div
  className="screen-watcher screen-gt-xs screen-gt-sm screen-md css-1ow008w"
/>
`;

exports[`ScreenWatcher renders the Component correctly for sm size 1`] = `
<div
  className="screen-watcher screen-gt-xs screen-sm css-1ow008w"
/>
`;

exports[`ScreenWatcher renders the Component correctly for xl size 1`] = `
<div
  className="screen-watcher screen-gt-xs screen-gt-sm screen-gt-md screen-gt-lg screen-xl css-1ow008w"
/>
`;

exports[`ScreenWatcher renders the Component correctly for xs size 1`] = `
<div
  className="screen-watcher screen-xs css-1ow008w"
/>
`;

exports[`ScreenWatcher renders the Container correctly 1`] = `
<div
  className="screen-watcher screen-gt-xs screen-gt-sm screen-md css-1ow008w"
/>
`;
